lib/scholar/metrics.ex

```diff
@@ -336,6 +336,93 @@ defmodule Scholar.Metrics do
+
+  ## Options
+
+    * `:num_classes` - required. Number of classes contained in the input tensors
+    * `:average` - optional. This determines the type of averaging performed on the data.
+      * `:macro`. Calculate metrics for each label, and find their unweighted mean.
+        This does not take label imbalance into account.
+      * `:weighted`. Calculate metrics for each label, and find their average weighted by
+        support (the number of true instances for each label).
+      * `:micro`. Calculate metrics globally by counting the total true positives,
+        false negatives and false positives.
+      * If not specified, then the f1 scores for each class are returned.
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: nil)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 0.6666666865348816, 0.4000000059604645]
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :macro)
+      #Nx.Tensor<
+        f32
+        0.5777778029441833
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :weighted)
+      #Nx.Tensor<
+        f32
+        0.64000004529953
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :micro)
+      #Nx.Tensor<
+        f32
+        0.6000000238418579
+      >
+  """
+  defn f1_score(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes, :average])
+
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    num_classes =
+      transform(opts[:num_classes], fn num_classes ->
+        num_classes || raise ArgumentError, "missing option :num_classes"
+      end)
+
+    average =
+      transform(opts[:average], fn average ->
+        average || nil
+      end)
+
+    cm = confusion_matrix(y_true, y_pred, num_classes: num_classes)
+    true_positive = Nx.take_diagonal(cm)
+    false_positive = Nx.subtract(Nx.sum(cm, axes: [0]), true_positive)
+    false_negative = Nx.subtract(Nx.sum(cm, axes: [1]), true_positive)
+    precision = Nx.divide(true_positive, true_positive + false_positive + 1.0e-16)
+    recall = Nx.divide(true_positive, true_positive + false_negative + 1.0e-16)
+
+    per_class_f1 =
+      Nx.divide(Nx.multiply(2, Nx.multiply(precision, recall)), precision + recall + 1.0e-16)
+
+    transform(average, fn average ->
+      case average do
+        nil ->
+          per_class_f1
+
+        :macro ->
+          Nx.mean(per_class_f1)
+
+        :weighted ->
+          support = Nx.sum(Nx.equal(y_true, Nx.iota({num_classes, 1})), axes: [1])
+          Nx.sum(Nx.multiply(per_class_f1, Nx.divide(support, Nx.sum(support) + 1.0e-16)))
+
+        :micro ->
+          y_pred
+          |> Nx.equal(y_true)
+          |> Nx.mean()
```

Comment:
This is just accuracy, is this correct?

Reply:
yes, but I got errors when I tried to use `if`, no matter inside a transform or not... Not sure how to do this correctly. 

btw, using `case` works, is that what I should use?

### `case` works
```
  defn f1_score(y_true, y_pred, opts \\ []) do
    opts = keyword!(opts, [:num_classes, :average])

    assert_shape_pattern(y_true, {_})
    assert_shape(y_pred, Nx.shape(y_true))

    transform(opts[:average], fn average ->
      case average do
        :micro ->
          accuracy(y_true, y_pred)

        _ ->
          # ...
          case average do
            nil ->
              # ...
            :macro ->
              # ...
            :weighted ->
              # ...
          end
      end
    end)
  end
```

### `if` does not work
```
  defn f1_score(y_true, y_pred, opts \\ []) do
    opts = keyword!(opts, [:num_classes, :average])

    assert_shape_pattern(y_true, {_})
    assert_shape(y_pred, Nx.shape(y_true))

    if opts[:average] == :micro do
        # ...
    else
        # ...
    end
  end

(ArgumentError) expected a %Nx.Tensor{} or a number, got: nil
```

```
  defn f1_score(y_true, y_pred, opts \\ []) do
    opts = keyword!(opts, [:num_classes, :average])

    assert_shape_pattern(y_true, {_})
    assert_shape(y_pred, Nx.shape(y_true))

    transform(opts[:average], fn average ->
      if average == :micro do
          # ...
      else
          # ...
      end
    end)
  end

(ArgumentError) expected a %Nx.Tensor{} or a number, got: nil
```

```
  defn f1_score(y_true, y_pred, opts \\ []) do
    opts = keyword!(opts, [:num_classes, :average])

    assert_shape_pattern(y_true, {_})
    assert_shape(y_pred, Nx.shape(y_true))

    average = transform(opts[:average], &(&1))
    if average == :micro do
        # ...
    else
        # ...
    end
  end

(ArgumentError) expected a %Nx.Tensor{} or a number, got: nil
```

Reply:
@cocoa-xu my question was rather whether f1 score `:micro` is just accuracy (?).

As for the error, inside transform the `==` is still the tensor-aware equality, you can do `Elixir.Kernel.==(average, :micro)` in those cases.

Reply:
yes, micro f1 is just the accuracy. (however, if we'd like to also implement [F-beta scores](https://en.wikipedia.org/wiki/F-score), then it's different.)

Reply:
Ah, I see, it's the same for single-label case! `fbeta` should be separate.